### PR TITLE
feat: [PL-31642]: Add retry logic for list & delete operations

### PR DESCRIPTION
**Pull request by mdmilic**
## Describe your changes
* Added createIfNotExists logic for Service
* Added retry logic for list & delegate operations for Pod, Secret & Service
* Added tests for creteIfNotExists for all resources
* Added tests for list & delete operations
* general refactoring

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49947)

### CODE REVIEW

**Comment by mdmilic** [comment]
trigger uts

File: 260-delegate/src/main/java/io/harness/delegate/service/core/apiclient/CoreV1ApiWithRetry.java [hunk 1]
```diff
@@ -7,105 +7,273 @@
...
     super(apiClient);
+    if (delay == null || delay.isNegative()) {
+      throw new IllegalArgumentException("Delay must be a positive value.");
+    }
+    this.delay = delay;
+  }
+
+  @Override
+  public V1PodList listNamespacedPod(final String namespace, final String pretty, final Boolean allowWatchBookmarks,
+      final String _continue, final String fieldSelector, final String labelSelector, final Integer limit,
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/delegate/service/core/apiclient/CoreV1ApiWithRetry.java:90]: is _continue a pagination param? that's why it's _ prefixed

**Review by xingchi-jin — APPROVED** [review]
with some nits

File: 260-delegate/src/main/java/io/harness/delegate/service/core/apiclient/CoreV1ApiWithRetry.java (hunk 1 above)
Comment (mdmilic) [reply 260-delegate/src/main/java/io/harness/delegate/service/core/apiclient/CoreV1ApiWithRetry.java:90]: Those are all the original names of the params as in the base class (k8s client)

File: 260-delegate/src/main/java/io/harness/delegate/service/core/apiclient/CoreV1ApiWithRetry.java [hunk 2]
```diff
@@ -7,105 +7,273 @@
...
-      return super.createNamespacedPod(namespace, body, pretty, dryRun, fieldManager, fieldValidation);
+      return super.createNamespacedSecret(namespace, body, pretty, dryRun, fieldManager, fieldValidation);
     } catch (final ApiException e) {
       if (e.getCode() != HTTP_CONFLICT) {
         throw e;
       }
-      log.warn("Pod {} already exists, replacing it", body.getMetadata().getName());
-      super.deleteNamespacedPod(body.getMetadata().getName(), namespace, pretty, dryRun, 0, null, null, null);
-      return super.createNamespacedPod(namespace, body, pretty, dryRun, fieldManager, fieldValidation);
+      log.warn("Secret1 {} already exists, replacing it", body.getMetadata().getName());
```
[the remark this answers is not in the document]
Comment (mdmilic) [reply 260-delegate/src/main/java/io/harness/delegate/service/core/apiclient/CoreV1ApiWithRetry.java:219]: will fix

**Pull request merged**